src/NuGet.Core/NuGet.Configuration/Settings/Items/FileClientCertItem.cs

```diff
@@ -0,0 +1,233 @@
+        {
+            get
+            {
+                string result = null;
+                if (IsPasswordIsClearText)
+                {
+                    Attributes.TryGetValue(ConfigurationConstants.ClearTextPasswordAttribute, out result);
+                }
+                else
+                {
+                    if (Attributes.TryGetValue(ConfigurationConstants.PasswordAttribute, out var encryptedPassword))
+                    {
+                        try
+                        {
+                            result = EncryptionUtility.DecryptString(encryptedPassword);
+                        }
+                        catch (Exception e)
+                        {
+                            throw new NuGetConfigurationException(string.Format(Resources.FileCertItemPasswordCannotBeDecrypted, PackageSource), e);
+                        }
+                    }
+                }
+
+                return result;
+            }
+        }
+
+        protected override IReadOnlyCollection<string> AllowedAttributes { get; } = new HashSet<string>
+        {
+            ConfigurationConstants.PackageSourceAttribute,
+            ConfigurationConstants.PathAttribute,
+            ConfigurationConstants.PasswordAttribute,
+            ConfigurationConstants.ClearTextPasswordAttribute
+        };
+
+        protected override IReadOnlyCollection<string> RequiredAttributes { get; } = new HashSet<string>
+        {
+            ConfigurationConstants.PackageSourceAttribute,
+            ConfigurationConstants.PathAttribute
+        };
+
+        internal override XNode AsXNode()
+        {
+            if (Node is XElement)
+            {
+                return Node;
+            }
+
+            var element = new XElement(ElementName);
+            foreach (KeyValuePair<string, string> attr in Attributes)
+            {
+                element.SetAttributeValue(attr.Key, attr.Value);
+            }
+
+            return element;
+        }
+
+        public override SettingBase Clone()
+        {
+            return new FileClientCertItem(PackageSource, FilePath, Password, IsPasswordIsClearText, Origin);
+        }
+
+        public override IEnumerable<X509Certificate> Search()
+        {
+            byte[] certificateData;
+
+            var filePath = FindAbsoluteFilePath();
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                throw new NuGetConfigurationException(string.Format(CultureInfo.CurrentCulture,
+                                                                    Resources.FileCertItemPathFileNotExist));
+            }
+
+            //Read certificate from file
+            using (FileStream stream = File.OpenRead(filePath))
+            {
+                certificateData = ReadStream(stream);
+            }
+
+            return new[] { string.IsNullOrWhiteSpace(Password) ? new X509Certificate2(certificateData) : new X509Certificate2(certificateData, Password) };
```

Comment:
Have you tried using the `X509Certificate2` constructor that can load from a file?  My concern is a large certificate might result in `ReadStream` allocating one or more objects in the Large Object Heap, which is not good for performance. Using methods in the BCL will hopefully have better performance. I think we can get rid of the `ReadStream` helper method.

Reply:
You are right. Usage was updated